### PR TITLE
Partition selection cleanups

### PR DESCRIPTION
**Pull request by hlinnaka**
See individual commits.

I came up with these, while reviewing PR #1462 and trying to understand how partition selectors work. These changes are in the same area as that PR, but shouldn't conflict too badly.

### CODE REVIEW

**Comment by vraghavan78** [comment]
@foyzur @armenatzoglou @karthijrk please take a look.

File: src/backend/executor/execUtils.c [hunk 1]
```diff
@@ -352,9 +329,13 @@ FreeExecutorState(EState *estate)
 		 * dynamicTableScanInfo != estate->dynamicTableScanInfo [JIRA: MPP-23562]
 		 */
 		Assert(proc_exit_inprogress || dynamicTableScanInfo != estate->dynamicTableScanInfo);
-		freeDynamicTableScanInfo(estate->dynamicTableScanInfo);
+		pfree(estate->dynamicTableScanInfo);
 		estate->dynamicTableScanInfo = NULL;
 	}
+	if (estate->partsMetadata != NIL)
+	{
+		list_free_deep(estate->partsMetadata);
```
Comment (foyzur) [inline src/backend/executor/execUtils.c]: Looks like we are leaking partsMetadata->accessMethods. We were already leaking before your changes though.

**Comment by foyzur** [comment]
Looks really good. Excellent refactoring :-) Thanks @hlinnaka for cleaning up a lot of the clutter.

**Comment by foyzur** [comment]
BTW @hlinnaka could you please check why concourse is read? For some weird reason I cannot seem to open the concourse build to check myself.

**Comment by hlinnaka** [comment]
I'm not sure what Concourse's problem was. I thought I saw it pass earlier, but the status was still "pending" in github. In any case, I've rebased these patches over current master. That included the changes in PR #1462, since you pushed that already. Let's see what Concourse thinks of it now. (I wish we had a test case with a query that has two partition selectors for a single scan in the regression suite.)

**Comment by cjcjameson** [comment]
@tom-meyer @schubert can we look into why the concourse status doesn't show up correctly in the PRs? We looked at it briefly yesterday and didn't come to a resolution.

**Comment by armenatzoglou** [comment]
Other than the change that @karthijrk recommended, it LGTM.

**Comment by hlinnaka** [comment]
The regression tests were not happy with this, unfortunately. It seems that there's a case where we have a plan, with a "static selector", where the scanId is not included in the numSelectorsPerScanId list. The current code copes with it, by dynamically extending the array in InsertPidIntoDynamicTableScanInfo(), if it's given a scanId that's not covered by the current array, but with these refactorings, I assumed that the array can be allocated at the beginning of the executor, to the size of numSelectorsPerScanId(). Having a partition selector, with a scanId with no entry in the numSelectorsPerScanId list seems bogus to me, but this clearly needs more investigation.

To move forward, I removed the third commit from this PR, which was the biggest refactoring, and caused that issue. I will push the rest, assuming that Concourse is happy with those alone, and open a new PR for the third one later, after investigating that failure.

**Comment by hlinnaka** [comment]
Pushed the three patches that passed regression tests, thanks!

**Pull request closed without merging**